### PR TITLE
Bugfix #21008, controller stage misbehaving on early response

### LINKED ISSUE #21008: Cannot pull port (requestParsingIn) twice
Environment: Java 8, scala 2.11.7, akka 2.4.8

Testing with:
curl -v -F"a=@test1.dat" -F"b=@test2.dat" http://localhost:7777/test

The following code results in a Timeout Error:

```
object AkkaTest extends App {
  implicit val system = ActorSystem()
  implicit val materializer = ActorMaterializer()
  implicit val executor = system.dispatcher

  val route = encodeResponse {
    (post & path("test") & entity(as[Multipart.FormData])) { entity =>
      complete {
        val sources = entity.parts.map(e => e.entity.dataBytes)
        val f = sources.runForeach { source =>
          val inputStream = source.runWith(StreamConverters.asInputStream())
          try {
            val tmp = Files.createTempFile("a", "b")
            Files.copy(inputStream, tmp, StandardCopyOption.REPLACE_EXISTING)
            Files.delete(tmp)
          } finally {
            inputStream.close()
          }
        }
        f.map(_ => "ok")
      }
    }
  }
  import akka.http.scaladsl.Http
  Http().bindAndHandle(route, "0.0.0.0", 7777)
```

Whereas, this works as expected (using runFold and ByteArrrayInputStream):

```
object AkkaTest extends App {
  implicit val system = ActorSystem()
  implicit val materializer = ActorMaterializer()
  implicit val executor = system.dispatcher

  val route = encodeResponse {
    (post & path("test") & entity(as[Multipart.FormData])) { entity =>
      complete {
        val sources = entity.parts.map(e => e.entity.dataBytes)
        val f = sources.runForeach { source =>
          val tmp = Files.createTempFile("a", "b")
          source.runFold(ByteString.empty)((bs, c) => bs ++ c)
            .foreach { bs =>
              val inputStream = new ByteArrayInputStream(bs.toArray)
              try {
                Files.copy(inputStream, tmp, StandardCopyOption.REPLACE_EXISTING)
                Files.delete(tmp)
              } finally {
                inputStream.close()
              }
            }
        }
        f.map(_ => "ok")
      }
    }
  }
  import akka.http.scaladsl.Http
  Http().bindAndHandle(route, "0.0.0.0", 7777)
}
```

Exception:
```
java.io.IOException: Timeout on waiting for new data
	at akka.stream.impl.io.InputStreamAdapter$$anonfun$read$1.apply$mcI$sp(InputStreamSinkStage.scala:144) ~[akka-stream_2.11-2.4.8.jar:na]
	at akka.stream.impl.io.InputStreamAdapter$$anonfun$read$1.apply(InputStreamSinkStage.scala:129) ~[akka-stream_2.11-2.4.8.jar:na]
	at akka.stream.impl.io.InputStreamAdapter$$anonfun$read$1.apply(InputStreamSinkStage.scala:129) ~[akka-stream_2.11-2.4.8.jar:na]
	at akka.stream.impl.io.InputStreamAdapter.executeIfNotClosed(InputStreamSinkStage.scala:109) ~[akka-stream_2.11-2.4.8.jar:na]
	at akka.stream.impl.io.InputStreamAdapter.read(InputStreamSinkStage.scala:129) ~[akka-stream_2.11-2.4.8.jar:na]
	at akka.stream.impl.io.InputStreamAdapter.read(InputStreamSinkStage.scala:120) ~[akka-stream_2.11-2.4.8.jar:na]
	at java.nio.file.Files.copy(Files.java:2908) ~[na:1.8.0_45]
	at java.nio.file.Files.copy(Files.java:3027) ~[na:1.8.0_45]
	at com.scb.sabre.store.http.AkkaTest$$anonfun$15$$anonfun$apply$69$$anonfun$apply$70$$anonfun$17.apply(
....
2016-07-20 10:02:13,202 - [ERROR] - [akka.actor.RepointableActorRef] [undefined] [default-akka.actor.default-dispatcher-2] - Error in stage [akka.http.impl.engine.server.HttpServerBluePrint$ControllerStage@5455fb41]: requirement failed: Cannot pull port (requestParsingIn) twice
java.lang.IllegalArgumentException: requirement failed: Cannot pull port (requestParsingIn) twice
	at scala.Predef$.require(Predef.scala:219) ~[scala-library-2.11.6.jar:na]
	at akka.stream.stage.GraphStageLogic.pull(GraphStage.scala:348) ~[akka-stream_2.11-2.4.8.jar:na]
	at akka.http.impl.engine.server.HttpServerBluePrint$ControllerStage$$anon$12$$anon$15.onPush(HttpServerBluePrint.scala:428) ~[akka-http-core_2.11-2.4.8.jar:na]
	at akka.stream.impl.fusing.GraphInterpreter.processElement$1(GraphInterpreter.scala:590) ~[akka-stream_2.11-2.4.8.jar:na]
	at akka.stream.impl.fusing.GraphInterpreter.processEvent(GraphInterpreter.scala:601) ~[akka-stream_2.11-2.4.8.jar:na]
	at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:542) ~[akka-stream_2.11-2.4.8.jar:na]
	at akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:471) ~[akka-stream_2.11-2.4.8.jar:na]

```

Possibly related to:
https://github.com/akka/akka/issues/19392
https://github.com/akka/akka/issues/20642

### PR DESCRIPTION
**Pull request by johanandren**
Fixes #21008

### CODE REVIEW

**Comment by johanandren** [comment]
Didn't manage to create a test to cover it. Found one pull on closed in and one push on already pushed out that can happen (very racey) when you complete a request early and try to consume its entities streamingly afterwards.

**Comment by drewhk** [comment]
Hm, would be nice to have tests though.

**Comment by ktoso** [comment]
LGTM though eagerly awaiting a test case for it :)

File: akka-http-core/src/test/scala/akka/http/impl/engine/server/HttpServerBug2108Spec.scala [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+              // and then when the failure has happened/future completes, we push a reply
+              responses.sendNext(HttpResponse(entity = "Yeah"))
+
+            }
+            // got such an error, that is bad,
+            sawException = true
+          } catch {
+            case _: AssertionError ⇒ sawException = false
+          }
+          if (sawException) fail("HttpServerBluePrint.ControllerStage: requirement failed: Cannot pull closed port (requestParsingIn)")
```
Comment (johanandren) [inline akka-http-core/src/test/scala/akka/http/impl/engine/server/HttpServerBug2108Spec.scala]: this is horrible but couldn't figure out a different way to do it

File: akka-http-core/src/test/scala/akka/http/impl/engine/server/HttpServerBug2108Spec.scala [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                 |
+                 |""")
+
+          // Bug #21008 does not actually ever make the request fail instead it logs the exception and behaves
+          // nicely so we need to check that the exception didn't get logged
+          var sawException = false
+          try {
+            EventFilter[IllegalArgumentException](occurrences = 1) intercept {
+              // make sure the failure has happened
+              Await.ready(done, 10.seconds)
```
Comment (ktoso) [inline akka-http-core/src/test/scala/akka/http/impl/engine/server/HttpServerBug2108Spec.scala]: would it still be ok with 5 seconds maybe?

**Comment by ktoso** [comment]
LGTM, no fun test but I don't have a better idea

File: akka-http-core/src/test/scala/akka/http/impl/engine/server/HttpServerBug2108Spec.scala (hunk 2 above)
Comment (johanandren) [inline akka-http-core/src/test/scala/akka/http/impl/engine/server/HttpServerBug2108Spec.scala]: It's not really an interesting timeout, will always complete quickly as the exception is thrown as soon as the server receives next chunk.

Comment (ktoso) [inline akka-http-core/src/test/scala/akka/http/impl/engine/server/HttpServerBug2108Spec.scala]: I see, then I don't mind the timeout I guess

**Pull request merged**